lib/em-hiredis/new_client.rb

```diff
@@ -0,0 +1,239 @@
+require 'uri'
+
+module EventMachine::Hiredis
+  # Emits the following events
+  #
+  # * :connected - on successful connection or reconnection
+  # * :reconnected - on successful reconnection
+  # * :disconnected - no longer connected, when previously in connected state
+  # * :reconnect_failed(failure_number) - a reconnect attempt failed
+  #     This event is passed number of failures so far (1,2,3...)
+  class BaseClient
+    include EventEmitter
+    include EventMachine::Deferrable
+
+    attr_reader :host, :port, :password, :db
+
+    TRANSITIONS = [
+      [ :connect,                 :initial, :connecting ],
+      [ :connect_failure,         :connecting, :connect_failed ],
+      [ :retry_connect,           :connect_failed, :connecting ],
+      [ :connect_perm_failure,    :connect_failed, :failed ],
+      [ :setup,                   :connecting, :initialising ],
+      [ :setup_failure,           :initialising, :setup_failed ],
+      [ :retry_setup,             :setup_failed, :connecting ],
+      [ :setup_perm_failure,      :setup_failed, :failed ],
+      [ :setup_success,           :initialising, :connected ],
+      [ :disconnected,            :connected, :disconnected ],
+      [ :reconnect,               :disconnected, :connecting ],
+      [ :recover,                 :failed, :connecting ],
+    ]
+
+    def initialize(uri)
+      configure(uri)
+
+      @reconnect_attempt = 0
+
+      # Commands received while we are not initialized, to be sent once we are
+      @command_queue = []
```

Comment:
If the EM::Connection just buffers incoming connections then that mechanism is not necessary isn't it ? If the EM::Connection would fail then all the callbacks would be failed automatically too.